translation(JS): web/javascript/reference/iteration_protocols (#1219)
webdoky/content: files/uk/web/javascript/reference/iteration_protocols/index.md

## Протокол ітерованого об'єкта

**Протокол ітерованого об'єкта** дає об'єктам JavaScript змогу означити або налаштувати власну логіку ітерування, наприклад, те, які значення обробляються в циклі {{jsxref("Statements/for...of", "for...of")}}. Частина вбудованих типів є [вбудованими ітерованими об'єктами](#vbudovani-iterovani-obiekty) з усталеною логікою ітерування, як то {{jsxref("Array")}} і {{jsxref("Map")}}, коли інші (як то {{jsxref("Object")}}) – ні.

Аби бути **ітерованим**, об'єкт мусить реалізувати метод **`@@iterator`**, що означає, що об'єкт (або один з об'єктів у його [ланцюжку прототипів](/uk/docs/Web/JavaScript/Inheritance_and_the_prototype_chain)) мусить мати властивість з ключем `@@iterator`, доступним через сталу {{jsxref("Symbol.iterator")}}:

- `[Symbol.iterator]`
  - : Функція без аргументів, що повертає об'єкт, котрий виконує [протокол ітератора](#protokol-iteratora).

Щоразу, коли об'єкт готується до ітерування (наприклад, на початку циклу {{jsxref("Statements/for...of", "for...of")}}), викликається без аргументів його метод `@@iterator`, і повернений **ітератор** використовується для отримання значень ітерування.

> Зверніть увагу, що коли викликається ця функція без аргументів, вона закликається як метод ітерованого об'єкта. Таким чином, всередині цієї функції для звертання до властивостей ітерованого об'єкта може використовуватися ключове слово `this`, для з'ясування того, що повинно видаватися під час ітерації.

Reviewer: для-для
```suggestion
Зверніть увагу, що коли викликається ця функція без аргументів, вона закликається як метод ітерованого об'єкта. Таким чином, всередині цієї функції для звертання до властивостей ітерованого об'єкта може використовуватися ключове слово `this`, аби з'ясувати, що повинно видаватися під час ітерації.
```

Ця функція може бути звичайною функцією, а може бути генераторною функцією, щоб при її заклику повертався об'єкт-ітератор. Всередині такої генераторної функції кожен запис надається за допомогою `yield`.

## Протокол ітератора

> **Протокол ітератора** задає стандартний спосіб вироблення послідовності значень (або скінченний, або нескінченний), і потенційно повернене значення, коли всі елементи послідовності були згенеровані.

Reviewer: ```suggestion
**Протокол ітератора** задає стандартний спосіб вироблення послідовності (або скінченної, або нескінченної) значень, і, потенційно, повернене значення, коли всі елементи послідовності були згенеровані.
```

Об'єкт є ітератором, коли має реалізацію метода **`next()`** з наступною семантикою:

- `next()`

>   - : Функція, що приймає нуль або один аргумент і повертає об'єкт, що відповідає інтерфейсові `IteratorResult` (див. нижче). Якщо повертається необ'єктне значення (наприклад, `false` або `undefined`), коли ітератор використовує вбудована можливість мови (як то `for...of`), то викидається {{jsxref("TypeError")}} (`"iterator.next() returned a non-object value"`).

Reviewer: Тут не одразу зрозуміло, що таке цікаве використовує цей ітератор, переформулював для однозначності
```suggestion
  - : Функція, що приймає нуль або один аргумент і повертає об'єкт, що відповідає інтерфейсові `IteratorResult` (див. нижче). Якщо повертається необ'єктне значення (наприклад, `false` або `undefined`), коли ітератор використовується вбудованою можливістю мови (як то `for...of`), то викидається {{jsxref("TypeError")}} (`"iterator.next() returned a non-object value"`).
```

- `return(value)` {{optional_inline}}
  - : Функція, котра приймає нуль або один аргумент і повертає об'єкт, котрий відповідає інтерфейсові `IteratorResult`, зазвичай із полем `value`, рівним переданому в нього `value`, і полем `done` зі значенням `true`. Виклик цього метода каже ітераторові, що викликач не має наміру більше викликати `next()` і що можна зайнятися очищенням.
- `throw(exception)` {{optional_inline}}
  - : Функція, котра приймає нуль або один аргумент і повертає об'єкт, що відповідає інтерфейсові `IteratorResult`, зазвичай із полем `done`, рівним `true`. Виклик цього метода каже ітераторові, що викликач увійшов у помилковий стан, а `exception` зазвичай є примірником {{jsxref("Error")}}.

> **Примітка:** Неможливо рефлективно дізнатися (тобто без фактичного виклику `next()` і перевірки поверненого результату), чи реалізує певний об'єкт протокол ітератора.

Дуже легко зробити ітератор також ітерованим об'єктом: достатньо реалізувати метод `[@@iterator]()`, котрий повертає `this`.

```js
// Відповідає і протоколові ітератора, і протоколові ітерованого об'єкта
const myIterator = {
  next() {
    // ...
  },
  [Symbol.iterator]() {
    return this;
  },
};
```

> Такий об'єкт зветься _ітерованим ітератором_. Це дає змогу використати ітератор в певних записах, що розраховують на ітеровані об'єкти – таким чином, нечасто є корисною реалізація протоколу ітератора без реалізації водночас протоколу ітерованого об'єкта. (Насправді майже всі записи й API очікують на _ітеровані об'єкти_, а не _ітератори_.) [Генераторний об'єкт](/uk/docs/Web/JavaScript/Reference/Global_Objects/Generator) є прикладом цього:

Reviewer: просто "запис" звучить трохи розмито, пропоную сформулювати ближче до оригіналу
```suggestion
Такий об'єкт зветься _ітерованим ітератором_. Це дає змогу використати ітератор в певних синтаксичних структурах, що розраховують на ітеровані об'єкти – таким чином, нечасто є корисною реалізація протоколу ітератора без реалізації водночас протоколу ітерованого об'єкта. (Насправді майже всі мовні структури й API очікують на _ітеровані об'єкти_, а не _ітератори_.) [Генераторний об'єкт](/uk/docs/Web/JavaScript/Reference/Global_Objects/Generator) є прикладом цього:
```

```js
const aGeneratorObject = (function* () {
  yield 1;
  yield 2;
  yield 3;
})();

console.log(typeof aGeneratorObject.next);
// "function": є метод next (котрий повертає правильний результат), тож це ітератор

console.log(typeof aGeneratorObject[Symbol.iterator]);
// "function": є метод @@iterator (котрий повертає правильний ітератор), тож це ітерований об'єкт

console.log(aGeneratorObject[Symbol.iterator]() === aGeneratorObject);
// true: метод @@iterator повертає сам об'єкт (сам ітератор), тож це ітерований ітератор
```

А проте, коли це можливо, краще, щоб `iterable[Symbol.iterator]` повертав різні ітератори, що завжди починаються спочатку, як це робить [`Set.prototype[@@iterator]()`](/uk/docs/Web/JavaScript/Reference/Global_Objects/Set/@@iterator).

## Асинхронний ітератор і протокол асинхронного ітерованого об'єкта

> Є іще одна пара протоколів – вона використовується для асинхронного ітерування, вони звуться протоколами **асинхронного ітератора** й **асинхронного ітерованого об'єкта**. Їх інтерфейси дуже схожі на інтерфейси протоколів ітерованого об'єкта та протокол ітератора, окрім того, що кожне повернене при викликах методів ітератора значення – загорнуте в проміс.

Reviewer: вона-вони
```suggestion
Є іще одна пара протоколів, яка використовується для асинхронного ітерування – вони звуться протоколами **асинхронного ітератора** й **асинхронного ітерованого об'єкта**. Їх інтерфейси дуже схожі на інтерфейси протоколів ітерованого об'єкта та ітератора, окрім того, що кожне повернене при викликах методів ітератора значення – загорнуте в проміс.
```

[Генераторні функції](/uk/docs/Web/JavaScript/Reference/Statements/function*) повертають [генераторні об'єкти](/uk/docs/Web/JavaScript/Reference/Global_Objects/Generator), котрі є ітерованими ітераторами. [Асинхронні генераторні функції](/uk/docs/Web/JavaScript/Reference/Statements/async_function*) повертають [асинхронні генераторні об'єкти](/uk/docs/Web/JavaScript/Reference/Global_Objects/AsyncGenerator), котрі є асинхронними ітерованими ітераторами.

Ітератори, повернені зі вбудованих ітерованих об'єктів, успадковують від спільного класу (наразі прихованого), котрий має реалізацію вищезгаданого методу `[Symbol.iterator]() { return this; }`, що робить їх всіх ітерованими ітераторами. У майбутньому ці вбудовані ітератори можуть отримати додаткові [допоміжні методи](https://github.com/tc39/proposal-iterator-helpers), на додачу до методу `next()`, котрий вимагається протоколом ітератора. Ланцюжок прототипів ітератора можна дослідити шляхом виведення його в графічну консоль.

```
console.log([][Symbol.iterator]());

Array Iterator {}

>   [[Prototype]]: Array Iterator     ==> This is the prototype shared by all array iterators

Reviewer: ```suggestion
  [[Prototype]]: Array Iterator     ==> Цей прототип — спільний для всіх ітераторів-масивів
```

    next: ƒ next()
    Symbol(Symbol.toStringTag): "Array Iterator"

>     [[Prototype]]: Object           ==> This is the prototype shared by all built-in iterators

Reviewer: ```suggestion
    [[Prototype]]: Object           ==> Цей прототип — спільний для всіх вбудованих ітераторів
```

      Symbol(Symbol.iterator): ƒ [Symbol.iterator]()

>       [[Prototype]]: Object         ==> This is Object.prototype

Reviewer: ```suggestion
      [[Prototype]]: Object         ==> Це — Object.prototype
```

```

### Вбудовані API, що приймають ітератори

Є чимало API, що приймають ітеровані об'єкти. Серед прикладів:

- {{jsxref("Map/Map", "Map()")}}
- {{jsxref("WeakMap/WeakMap", "WeakMap()")}}
- {{jsxref("Set/Set", "Set()")}}
- {{jsxref("WeakSet/WeakSet", "WeakSet()")}}
- {{jsxref("Promise.all()")}}
- {{jsxref("Promise.allSettled()")}}
- {{jsxref("Promise.race()")}}
- {{jsxref("Promise.any()")}}
- {{jsxref("Array.from()")}}

```js
const myObj = {};

new WeakSet(
  (function* () {
    yield {};
    yield myObj;
    yield {};
  })()
).has(myObj); // true
```

> ### Записи синтаксису, що очікують на ітеровані об'єкти

Reviewer: Такий термін більш вживаний
```suggestion
### Синтаксичні конструкції, що очікують на ітеровані об'єкти
```

Частина інструкцій і виразів очікує на ітеровані об'єкти, наприклад, цикли {{jsxref("Statements/for...of", "for...of")}}, [розгортання масивів і параметрів](/uk/docs/Web/JavaScript/Reference/Operators/Spread_syntax), {{jsxref("Operators/yield*", "yield*")}} і [деструктурування масивів](/uk/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment):

```js
for (const value of ["a", "b", "c"]) {
  console.log(value);
}
// "a"
// "b"
// "c"

console.log([..."abc"]); // ["a", "b", "c"]

function* gen() {
  yield* ["a", "b", "c"];
}

console.log(gen().next()); // { value: "a", done: false }

[a, b, c] = new Set(["a", "b", "c"]);
console.log(a); // "a"
```

> Коли вбудовані записи ітерують ітератор, і поле `done` останнього результату дорівнює `false` (тобто ітератор може виробити більше значень), але більше значень не потрібно, то буде викликано метод `return`, якщо такий метод є. Це може статися, наприклад, якщо в циклі `for...of` зустрілася інструкція `break` або `return`, або коли при деструктуруванні масиву усі ідентифікатори вже отримали значення.

Reviewer: ```suggestion
Коли вбудовані синтаксичні конструкції ітерують ітератор, і поле `done` останнього результату дорівнює `false` (тобто ітератор може виробити більше значень), але більше значень не потрібно, то буде викликано метод `return`, якщо такий метод є. Це може статися, наприклад, якщо в циклі `for...of` зустрілася інструкція `break` або `return`, або коли при деструктуруванні масиву усі ідентифікатори вже отримали значення.
```
